File: frontend/src/models/websocket-message/websocket-message.ts
import { ActionType } from '../action/action';
import { Character } from '../card/card';
import { GameStatePublic, GameStatePrivate } from '../game/game';

// --- Client → Server ---

export enum ClientMessageType {
  ACTION = 'action',
  CHALLENGE = 'challenge',
  BLOCK = 'block',
  ACCEPT = 'accept',
  CHOOSE_INFLUENCE = 'choose_influence',
  EXCHANGE_RETURN = 'exchange_return',
}

export interface ClientActionPayload {
  actionType: ActionType;
  targetId?: string;
}

export interface ClientChallengePayload {
  targetPlayerId: string;
}

export interface ClientBlockPayload {
  blockingCharacter: Character;
}

export interface ClientChooseInfluencePayload {
  cardIndex: number;
}

export interface ClientExchangeReturnPayload {
  keepIndices: number[];
}

export interface ClientMessage {
  type: ClientMessageType;
  payload:
    | ClientActionPayload
    | ClientChallengePayload
    | ClientBlockPayload
    | ClientChooseInfluencePayload
    | ClientExchangeReturnPayload
    | Record<string, never>;
}

// --- Server → Client ---

export enum ServerMessageType {
  GAME_STATE = 'game_state',
  PRIVATE_STATE = 'private_state',
  ACTION_DECLARED = 'action_declared',
  CHALLENGE_ISSUED = 'challenge_issued',
  CHALLENGE_RESULT = 'challenge_result',
  BLOCK_DECLARED = 'block_declared',
  BLOCK_CHALLENGE_RESULT = 'block_challenge_result',
  ACTION_RESOLVED = 'action_resolved',
  INFLUENCE_LOST = 'influence_lost',
  PLAYER_ELIMINATED = 'player_eliminated',
  TURN_CHANGED = 'turn_changed',
  GAME_OVER = 'game_over',
  PLAYER_JOINED = 'player_joined',
  PLAYER_LEFT = 'player_left',
  ERROR = 'error',
  EXCHANGE_CARDS = 'exchange_cards',
}

export interface ServerMessage {
  type: ServerMessageType;
  payload: Record<string, unknown>;
  gameState?: GameStatePublic;
  privateState?: GameStatePrivate;
  timestamp: string;
}

File: backend/app/models/websocket_message.py
"""WebSocket message schemas — discriminated union for type safety."""

from __future__ import annotations

from enum import Enum
from typing import Any

from pydantic import BaseModel, ConfigDict


# ── Client → Server message types ────────────────────────────────────────────


class ClientMessageType(str, Enum):
    ACTION = "ACTION"
    CHALLENGE = "CHALLENGE"
    BLOCK = "BLOCK"
    ACCEPT = "ACCEPT"  # Accept / pass on challenge/block window
    CHOOSE_INFLUENCE = "CHOOSE_INFLUENCE"
    EXCHANGE_RETURN = "EXCHANGE_RETURN"


class ClientMessage(BaseModel):
    """Message from client to server."""

    model_config = ConfigDict(strict=True)

    type: ClientMessageType
    payload: dict[str, Any] = {}


# ── Server → Client message types ────────────────────────────────────────────


class ServerMessageType(str, Enum):
    GAME_STATE = "GAME_STATE"
    ACTION_DECLARED = "ACTION_DECLARED"
    CHALLENGE_ISSUED = "CHALLENGE_ISSUED"
    CHALLENGE_WINDOW_OPEN = "CHALLENGE_WINDOW_OPEN"
    BLOCK_WINDOW_OPEN = "BLOCK_WINDOW_OPEN"
    CHALLENGE_RESULT = "CHALLENGE_RESULT"
    BLOCK_DECLARED = "BLOCK_DECLARED"
    BLOCK_CHALLENGE_RESULT = "BLOCK_CHALLENGE_RESULT"
    ACTION_RESOLVED = "ACTION_RESOLVED"
    INFLUENCE_LOST = "INFLUENCE_LOST"
    EXCHANGE_STARTED = "EXCHANGE_STARTED"
    PLAYER_ELIMINATED = "PLAYER_ELIMINATED"
    TURN_CHANGED = "TURN_CHANGED"
    GAME_OVER = "GAME_OVER"
    ERROR = "ERROR"
    PLAYER_CONNECTED = "PLAYER_CONNECTED"
    PLAYER_DISCONNECTED = "PLAYER_DISCONNECTED"
    WAITING_FOR = "WAITING_FOR"


class ServerMessage(BaseModel):
    """Message from server to client."""

    model_config = ConfigDict(strict=True)

    type: ServerMessageType
    payload: dict[str, Any] = {}
